x-pack/legacy/plugins/infra/public/utils/infra_redirect.tsx

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React from 'react';
+import { Route, Redirect, RouteProps } from 'react-router-dom';
+
+// This workaround preserves query parameters in the redirect
+// https://github.com/ReactTraining/react-router/issues/5818#issuecomment-379212014
+export const InfraRedirect: React.FunctionComponent<{
+  from: string;
+  to: string;
+  exact?: boolean;
+}> = ({ from, to, exact }) => (
+  <Route
+    path={from}
+    component={({ location }: RouteProps) =>
+      location ? (
+        <Redirect
+          from={from}
+          exact={exact}
+          to={{
+            ...location,
+            pathname: location.pathname.replace(from, to),
```

Comment:
Wouldn't just

```suggestion
            pathname: to,
```

do the trick? We're not looking for partial path replacements, I think.

Reply:
That doesn't work with things like `:type/:node`. It will redirect to literally `/infrastructure/metrics/:type/:node` if we include that in the `to`, or not include them at all if we just keep it as a redirect from `/metrics` to `/infrastructure/metrics`.

Reply:
oh, that's weird... the docs explicitly state that the placeholders should work:

> to: object
>
> A location to redirect to. `pathname` can be any valid URL path that path-to-regexp@^1.7.0 understands.

Reply:
ah, I see... it only works when used as a direct child of `<Switch>`. I wonder if we can make that work somehow... :thinking:

Reply:
It seems to be as simple as just forwarding the `...rest` we received from the `<Switch>`:

```typescript
import React from 'react';
import { Route, Redirect, RedirectProps, RouteProps } from 'react-router-dom';

// This workaround preserves query parameters in the redirect
// https://github.com/ReactTraining/react-router/issues/5818#issuecomment-379212014
export const RedirectWithQueryParams: React.FunctionComponent<RedirectProps> = ({
  from,
  to,
  ...rest
}) => (
  <Route
    path={from}
    render={({ location }: RouteProps) =>
      location ? (
        <Redirect
          {...rest}
          from={from}
          to={
            typeof to === 'string'
              ? {
                  ...location,
                  pathname: to,
                }
              : {
                  ...location,
                  ...to,
                }
          }
        />
      ) : null
    }
  />
);

```